src/category.rs

```diff
@@ -0,0 +1,235 @@
+        let rows = try!(stmt.query(&[&slug]));
+        rows.iter().next()
+                   .chain_error(|| NotFound)
+                   .map(|row| Model::from_row(&row))
+    }
+
+    pub fn encodable(self) -> EncodableCategory {
+        let Category { id: _, crates_cnt, category, slug, created_at } = self;
+        EncodableCategory {
+            id: slug.clone(),
+            slug: slug.clone(),
+            created_at: ::encode_time(created_at),
+            crates_cnt: crates_cnt,
+            category: category,
+        }
+    }
+
+    pub fn update_crate(conn: &GenericConnection,
+                        krate: &Crate,
+                        categories: &[String]) -> CargoResult<Vec<String>> {
+        let old_categories = try!(krate.categories(conn));
+        let old_categories_ids: HashSet<_> = old_categories.iter().map(|cat| {
+            cat.id
+        }).collect();
+
+        // If a new category specified is not in the database, filter
+        // it out and don't add it. Return it to be able to warn about it.
+        let mut invalid_categories = vec![];
+        let new_categories: Vec<Category> = categories.iter().flat_map(|c| {
+            match Category::find_by_slug(conn, &c) {
+                Ok(cat) => Some(cat),
+                Err(_) => {
+                    invalid_categories.push(c.to_string());
+                    None
+                },
+            }
+        }).collect();
+
+        let new_categories_ids: HashSet<_> = new_categories.iter().map(|cat| {
+            cat.id
+        }).collect();
+
+        let to_rm: Vec<_> = old_categories_ids
+                                .difference(&new_categories_ids)
+                                .cloned()
+                                .collect();
+        let to_add: Vec<_> = new_categories_ids
+                                .difference(&old_categories_ids)
+                                .cloned()
+                                .collect();
+
+        if !to_rm.is_empty() {
+            try!(conn.execute("DELETE FROM crates_categories \
+                                WHERE category_id = ANY($1) \
+                                  AND crate_id = $2",
+                              &[&to_rm, &krate.id]));
+        }
+
+        if !to_add.is_empty() {
+            let insert: Vec<_> = to_add.into_iter().map(|id| {
+                format!("({}, {})", krate.id, id)
+            }).collect();
+            let insert = insert.join(", ");
+            try!(conn.execute(&format!("INSERT INTO crates_categories \
+                                        (crate_id, category_id) VALUES {}",
+                                       insert),
+                              &[]));
+        }
+
+        Ok(invalid_categories)
+    }
+
+    pub fn count_toplevel(conn: &GenericConnection) -> CargoResult<i64> {
+        let sql = format!("\
+            SELECT COUNT(*) \
+            FROM {} \
+            WHERE category NOT LIKE '%::%'",
```

Comment:
Are we thinking two levels of nesting max? I wonder if it's perhaps better to have a `parent_id` field in the table for a checked pointer?